devdocs/testing_guidelines.md

```diff
@@ -0,0 +1,92 @@
+
+## Unit Tests
+
+Unit tests are critical for fast feedback. They should cover the logic of CRUD operations by mocking out HTTP calls to the Power Platform APIs. All unit tests for a given resource or data source are located in the `/internal/<resource_or_datasource>_test.go` file. JSON for mocked API responses can often be captured using network traces from browser debugging tools. Utilizing features in the Power Platform Admin Center can provide examples of how the Power Platform UI interacts with these APIs. While a test-first approach is not mandatory, it can be beneficial, particularly when using mocked data.&#x20;
+
+- This provider uses the `httpmock` library to simulate API responses. This approach was chosen because it allows us to also test serialization and deserialization of API requests and responses.
+- Common mocks (e.g., creating a Power Platform Environment) are encapsulated in helper functions for reuse, such as the `ActivateEnvironmentHttpMocks` function.
+- When adding new unit tests, place static JSON responses in `internal/services/<service>/test/<resource_or_datasource>/<test_name>/...`.
+- **Do not include real personal data** – anonymize IDs, emails, tenant IDs, phone numbers, etc.
+- Cover **create, read, update, and delete** behaviors, along with conditional logic (e.g., forced recreation).
+- Include **negative tests** for API errors (e.g., 403 Forbidden), invalid inputs, and boundary conditions.
+- When creating mocked JSON responses, you can reuse existing ones by duplicating them into your `<test_name>` folder.
+
+### Running Unit Tests
+
+To run all unit tests:
+
+```bash
+make unittest
+```
+
+To run a single unit test:
+
+```bash
+TF_ACC=0 go test -v ./... -run TestUnit<test_name>
+```
+
+## Acceptance Tests (Integration Tests)
+
+Every unit test covering a new feature or fix should have a corresponding acceptance test validating the same use case against real infrastructure. Acceptance tests ensure provider correctness in actual Power Platform environments.
+
+- Acceptance tests (files with `TestAcc...`) call real APIs and require valid credentials and a test tenant.
+- **Tests create real resources** – ensure proper cleanup after execution.
+  - Use **CheckDestroy** to verify resource deletion post `terraform destroy`.
+
+
+
+### Test Pre-checks
+
+Acceptance tests require authentication and appropriate test environment setup. This section is particularly useful for testing different authentication styles, ensuring the provider supports various authentication mechanisms reliably.
+
+- The test suite includes **pre-checks** (`testAccPreCheck(t)`) to validate required environment variables before running tests.
+- Ensure necessary credentials are set before executing acceptance tests.
+- Missing variables will result in skipped tests to prevent erroneous failures.
+
+#### Running Acceptance Tests
+
+To run all acceptance tests:
+
+```bash
+make acctest
```

Comment:
I think you need "TF_ACC=1" to get them running? Or we have that set in the container already?

Reply:
Makefile handles that for unittest

unittest:
	clear
	$(MAKE) clean
	$(MAKE) install
	TF_ACC=0 go test -p 16 -timeout 10m -v ./... -run "^TestUnit$(TEST)"

acctest:
	clear
	$(MAKE) clean
	$(MAKE) install
	TF_ACC=1 go test -p 10 -timeout 300m -v ./... -run "^TestAcc$(TEST)"